source/gui/__init__.py

```diff
@@ -360,6 +356,20 @@ def onConfigProfilesCommand(self, evt):
 		self.postPopup()
 
 
+def safeAppExit():
+	"""
+	Ensures the app is exited by all the top windows being destroyed
+	"""
+
+	for window in wx.GetTopLevelWindows():
+		if isinstance(window, wx.Dialog):
+			if window.IsModal():
+				wx.CallAfter(window.EndModal, wx.ID_CLOSE_ALL)
+		if not isinstance(window, MainFrame):
+			wx.CallAfter(window.Close)
+		else:
+			wx.CallAfter(window.Destroy)
+

```

Comment:
It might be helpful to log what is being closed. I think closing modal windows is new, so a log message to explain might be good. For instance, in past builds if the 'about' window was open you couldn't exit NVDA. This would be a nice to thing to fix, and there is probably an issue for it.

Reply:
Fixed, here's an example log with a lot of windows open during the exit

```python
INFO - gui.safeAppExit (12:19:45.692) - MainThread (9512):
destroying main frame during exit process
INFO - gui.safeAppExit (12:19:45.693) - MainThread (9512):
closing window <wx._core.Frame object at 0x00FA6F58> during exit process
INFO - gui.safeAppExit (12:19:45.693) - MainThread (9512):
ending modal <gui.startupDialogs.WelcomeDialog object at 0x0F084340> during exit process
INFO - gui.safeAppExit (12:19:45.693) - MainThread (9512):
closing window <gui.startupDialogs.WelcomeDialog object at 0x0F084340> during exit process
INFO - gui.safeAppExit (12:19:45.693) - MainThread (9512):
closing window <gui.logViewer.LogViewer object at 0x00AB34A8> during exit process
INFO - gui.safeAppExit (12:19:45.693) - MainThread (9512):
closing window <pythonConsole.ConsoleUI object at 0x122EF028> during exit process
INFO - gui.safeAppExit (12:19:45.693) - MainThread (9512):
closing window <gui.settingsDialogs.NVDASettingsDialog object at 0x00FA94F0> during exit process
INFO - gui.safeAppExit (12:19:45.693) - MainThread (9512):
closing window <gui.ExitDialog object at 0x010F3388> during exit process
INFO - core.main (12:19:46.391) - MainThread (9512):
Exiting
INFO - config.ConfigManager.save (12:19:46.406) - MainThread (9512):
Base configuration saved
INFO - external:__main__ (12:19:47.513) - MainThread (9512):
NVDA exit
```

Reply:
I'll have a look for related issues as well this might close